content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
+
+We had two options: 
+
+1. Modify `IRData` to allow us to have `IRName` or `IRConst` or `IRMem` nodes at each element in the data array instead of `long`s. 
+2. Initialize a non-static dispatch vector for each object on construction so we could use `IRName`, `IRConst` and `IRMem` to represent the dispatch vector and offsets into it.
+
+Both of these approaches would mean there were no changes to the `interpet` function of the interpeter, would allow the generated IR to work for both the `IRSimulator` and for assembley generation. The first approach however requried new classes to be created and modification of the `initialize` function of the `IRSimulator` to work with the new data class.
+
+We opted for approach 2 for simplicity, although approach 2 is definetly the better approach. We discuss this more in future work section.
+
+# Future Work
+
+## Typechecking
+
+This project skipped over the type checking phase so that more time could be spent focusing on the actual object oriented features. Obviously, this allows Xi programs that are not well typed and would likley crash or misbehave if these types of programs are used. Additionally, type checking would allow programs that involve more than one class. At the moment, it is impossible to figure out what class a variable `o` belongs to in situations such as `o.f()` and `o.x`. Currently, they are hardcoded to assume that all objects are of type `A`. Type checking would allow the compiler to identify what type a variable has and therefore make decisions based on class layout for any class type.
+
+## Fields
+
+All fields are private to a class. This means that if a class `A` has field `x` and class `B` entends `A`, then class `B` cannot refer to `x`. However, `B` can call methods of `A` and those methods may refer to `x`. Currently, subclasses do not allocate fields that belong to their parent classes. Therefore when a method of a super class is called, it has no way to get those fields. In the current implementation it will try and get the variable from the expected index which will really be an
+undefined memory access.
+
+In order to remedy this problem, we must allocate fields belonging to super classes at runtime since fields are not part of class declarations and are only part of class definitions by the Xi++ specification. I spoke with Professer Myers about this. C++ solves this problem by requiring public and private fields to be declared in the header files (the class declaration) so this information is known at compile time. Java on the otherhand does this at runtime by having a bytecode instruction that refers to the variables by name. In Xi, the runtime would probably need to do something similiar to the dispatch vector indexing scheme where all variables of a subtype are placed at a certain index in the object layout at runtime.
```

Comment:
I don't quite understand this conclusion: why is it necessary to allocate space at run time? It seems like the "superclass chain" for every class is known statically, so you could just walk that and gather up all the fields for all the classes. You would need to lay out the data fields carefully so that a, if `A` inherits from `B`, then an `A` object can be safely accessed through a `B` reference… but that's just a matter of appending subclass fields at the end.

I also don't quite understand the distinction between definitions and declarations in this context:

> fields are not part of class declarations and are only part of class definitions by the Xi++ specification

…but maybe that's because I don't know Xi well enough.

Reply:
It is true that Java bytecode seems to contain field names, but it is not true that the compiled code actually dynamically uses that (string) name…